### PR TITLE
feat: Keep screen awake

### PR DESCRIPTION
**Pull request by yanguoyu**
Add an option for users to keep the screen awake. Refer to https://github.com/Magickbase/neuron-public-issues/issues/319

Because the screen recording cannot capture the screen lock, the video was not uploaded.

### CODE REVIEW

**Review by devchenyan — APPROVED** [review]

**Comment by Danie0918** [comment]
@homura Please have a review.

File: packages/neuron-ui/src/utils/wakeScreen.ts [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+
+export async function wakeScreen() {
+  try {
+    if (!wakeLock) {
+      wakeLock = await navigator.wakeLock.request('screen')
+      document.addEventListener('visibilitychange', reAwake)
+    }
+  } catch (err) {
+    console.error('')
+  }
```
Comment (homura) [inline packages/neuron-ui/src/utils/wakeScreen.ts:20]:
```suggestion
  if (wakeLock) return

  try {
    wakeLock = await navigator.wakeLock.request('screen')
    document.addEventListener('visibilitychange', reAwake)
  } catch (err) {
    // do nothing here
  }
```

early return for readability

**Comment by silySuper** [comment]
After about  1 hour Lock screen ，sync process is increased only about 1%；and sync can not change when unlock screen.
an error show in main.log


[error] Unhandled Rejection in task:	Reason: AbortError: This operation was aborted
    at Object.fetch (node:internal/deps/undici/undici:11372:11)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async LightRPC.getTransactions (/Users/chllp/Downloads/Neuron 3.app/Contents/Resources/app/node_modules/@ckb-lumos/rpc/lib/method.js:57:19)
    at async LightSynchronizer.getTransactions (/Users/chllp/Downloads/Neuron 3.app/Contents/Resources/app/dist/block-sync-renderer/sync/light-synchronizer.js:88:28)
    at async LightSynchronizer.upsertTxHashes (/Users/chllp/Downloads/Neuron 3.app/Contents/Resources/app/dist/block-sync-renderer/sync/light-synchronizer.js:104:29)
    at async LightSynchronizer.synchronize (/Users/chllp/Downloads/Neuron 3.app/Contents/Resources/app/dist/block-sync-renderer/sync/light-synchronizer.js:76:29)
    at async LightSynchronizer.initSync (/Users/chllp/Downloads/Neuron 3.app/Contents/Resources/app/dist/block-sync-renderer/sync/light-synchronizer.js:202:13)
[2024-04-10T11:13:24.972Z] [error] Sync:ChildProcess: [2024-04-10T11:13:24.967Z] [error] Unhandled Rejection in task:	Reason: AbortError: This operation was aborted
    at Object.fetch (node:internal/deps/undici/undici:11372:11)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async LightRPC.getTransactions (/Users/chllp/Downloads/Neuron 3.app/Contents/Resources/app/node_modules/@ckb-lumos/rpc/lib/method.js:57:19)
    at async LightSynchronizer.getTransactions (/Users/chllp/Downloads/Neuron 3.app/Contents/Resources/app/dist/block-sync-renderer/sync/light-synchronizer.js:88:28)
    at async LightSynchronizer.upsertTxHashes (/Users/chllp/Downloads/Neuron 3.app/Contents/Resources/app/dist/block-sync-renderer/sync/light-synchronizer.js:104:29)
    at async LightSynchronizer.synchronize (/Users/chllp/Downloads/Neuron 3.app/Contents/Resources/app/dist/block-sync-renderer/sync/light-synchronizer.js:76:29)
    at async LightSynchronizer.initSync (/Users/chllp/Downloads/Neuron 3.app/Contents/Resources/app/dist/block-sync-renderer/sync/light-synchronizer.js:202:13)


[main.log](https://github.com/nervosnetwork/neuron/files/14931975/main.log)

**Comment by yanguoyu** [comment]
> After about 1 hour Lock screen ，sync process is increased only about 1%；and sync can not change when unlock screen. an error show in main.log
> 
> [error] Unhandled Rejection in task: Reason: AbortError: This operation was aborted at Object.fetch (node:internal/deps/undici/undici:11372:11) at process.processTicksAndRejections (node:internal/process/task_queues:95:5) at async LightRPC.getTransactions (/Users/chllp/Downloads/Neuron 3.app/Contents/Resources/app/node_modules/@ckb-lumos/rpc/lib/method.js:57:19) at async LightSynchronizer.getTransactions (/Users/chllp/Downloads/Neuron 3.app/Contents/Resources/app/dist/block-sync-renderer/sync/light-synchronizer.js:88:28) at async LightSynchronizer.upsertTxHashes (/Users/chllp/Downloads/Neuron 3.app/Contents/Resources/app/dist/block-sync-renderer/sync/light-synchronizer.js:104:29) at async LightSynchronizer.synchronize (/Users/chllp/Downloads/Neuron 3.app/Contents/Resources/app/dist/block-sync-renderer/sync/light-synchronizer.js:76:29) at async LightSynchronizer.initSync (/Users/chllp/Downloads/Neuron 3.app/Contents/Resources/app/dist/block-sync-renderer/sync/light-synchronizer.js:202:13) [2024-04-10T11:13:24.972Z] [error] Sync:ChildProcess: [2024-04-10T11:13:24.967Z] [error] Unhandled Rejection in task: Reason: AbortError: This operation was aborted at Object.fetch (node:internal/deps/undici/undici:11372:11) at process.processTicksAndRejections (node:internal/process/task_queues:95:5) at async LightRPC.getTransactions (/Users/chllp/Downloads/Neuron 3.app/Contents/Resources/app/node_modules/@ckb-lumos/rpc/lib/method.js:57:19) at async LightSynchronizer.getTransactions (/Users/chllp/Downloads/Neuron 3.app/Contents/Resources/app/dist/block-sync-renderer/sync/light-synchronizer.js:88:28) at async LightSynchronizer.upsertTxHashes (/Users/chllp/Downloads/Neuron 3.app/Contents/Resources/app/dist/block-sync-renderer/sync/light-synchronizer.js:104:29) at async LightSynchronizer.synchronize (/Users/chllp/Downloads/Neuron 3.app/Contents/Resources/app/dist/block-sync-renderer/sync/light-synchronizer.js:76:29) at async LightSynchronizer.initSync (/Users/chllp/Downloads/Neuron 3.app/Contents/Resources/app/dist/block-sync-renderer/sync/light-synchronizer.js:202:13)
> 
> [main.log](https://github.com/nervosnetwork/neuron/files/14931975/main.log)

```
[2024-04-10T10:57:52.809Z] [info]  Database:	saved synced block #2110000
[2024-04-10T11:13:24.967Z] [error] Unhandled Rejection in task:	Reason: AbortError: This operation was aborted
    at Object.fetch (node:internal/deps/undici/undici:11372:11)
```

Hi, when is the window locked? From the log, after 15 minutes in normal logs, the error log shows. This may be not a problem with this PR, but I will still find out what happened.

**Comment by silySuper** [comment]
> > After about 1 hour Lock screen ，sync process is increased only about 1%；and sync can not change when unlock screen. an error show in main.log
> > [error] Unhandled Rejection in task: Reason: AbortError: This operation was aborted at Object.fetch (node:internal/deps/undici/undici:11372:11) at process.processTicksAndRejections (node:internal/process/task_queues:95:5) at async LightRPC.getTransactions (/Users/chllp/Downloads/Neuron 3.app/Contents/Resources/app/node_modules/@ckb-lumos/rpc/lib/method.js:57:19) at async LightSynchronizer.getTransactions (/Users/chllp/Downloads/Neuron 3.app/Contents/Resources/app/dist/block-sync-renderer/sync/light-synchronizer.js:88:28) at async LightSynchronizer.upsertTxHashes (/Users/chllp/Downloads/Neuron 3.app/Contents/Resources/app/dist/block-sync-renderer/sync/light-synchronizer.js:104:29) at async LightSynchronizer.synchronize (/Users/chllp/Downloads/Neuron 3.app/Contents/Resources/app/dist/block-sync-renderer/sync/light-synchronizer.js:76:29) at async LightSynchronizer.initSync (/Users/chllp/Downloads/Neuron 3.app/Contents/Resources/app/dist/block-sync-renderer/sync/light-synchronizer.js:202:13) [2024-04-10T11:13:24.972Z] [error] Sync:ChildProcess: [2024-04-10T11:13:24.967Z] [error] Unhandled Rejection in task: Reason: AbortError: This operation was aborted at Object.fetch (node:internal/deps/undici/undici:11372:11) at process.processTicksAndRejections (node:internal/process/task_queues:95:5) at async LightRPC.getTransactions (/Users/chllp/Downloads/Neuron 3.app/Contents/Resources/app/node_modules/@ckb-lumos/rpc/lib/method.js:57:19) at async LightSynchronizer.getTransactions (/Users/chllp/Downloads/Neuron 3.app/Contents/Resources/app/dist/block-sync-renderer/sync/light-synchronizer.js:88:28) at async LightSynchronizer.upsertTxHashes (/Users/chllp/Downloads/Neuron 3.app/Contents/Resources/app/dist/block-sync-renderer/sync/light-synchronizer.js:104:29) at async LightSynchronizer.synchronize (/Users/chllp/Downloads/Neuron 3.app/Contents/Resources/app/dist/block-sync-renderer/sync/light-synchronizer.js:76:29) at async LightSynchronizer.initSync (/Users/chllp/Downloads/Neuron 3.app/Contents/Resources/app/dist/block-sync-renderer/sync/light-synchronizer.js:202:13)
> > [main.log](https://github.com/nervosnetwork/neuron/files/14931975/main.log)
> 
> ```
> [2024-04-10T10:57:52.809Z] [info]  Database:	saved synced block #2110000
> [2024-04-10T11:13:24.967Z] [error] Unhandled Rejection in task:	Reason: AbortError: This operation was aborted
>     at Object.fetch (node:internal/deps/undici/undici:11372:11)
> ```
> 
> Hi, when is the window locked? From the log, after 15 minutes in normal logs, the error log shows. This may be not a problem with this PR, but I will still find out what happened.

From about 19：00  to 20:00 yesterday

**Comment by yanguoyu** [comment]
@silySuper  Do you set `the keep screen awake` on when the screen is locked?

**Comment by silySuper** [comment]
yes

**Comment by yanguoyu** [comment]
> yes

So, the actual problem is when you set the `keep the screen awake` on, the screen is still locked after the preference locked time ends? Have you ever minimized the Neuron window?

**Comment by silySuper** [comment]
> > yes
> 
> So, the actual problem is when you set the `keep the screen awake` on, the screen is still locked after the preference locked time ends? Have you ever minimized the Neuron window?

I can not remember whether I have minimized the Neuron window.Because I have not see any tips about Neuron window operation before,so I do not pay close attention to this point.I have a screen recording before locking screen.

https://github.com/nervosnetwork/neuron/assets/25971273/81ade056-570b-4f3d-a6fb-0e5c92d43f84

After half an hour:

https://github.com/nervosnetwork/neuron/assets/25971273/64f4f0b2-ae85-4464-a590-504d3c8c5b57


[main.log](https://github.com/nervosnetwork/neuron/files/14942224/main.log)


And after half an hour I will offer a sync  process picture.

**Comment by yanguoyu** [comment]
> > > yes
> > 
> > 
> > So, the actual problem is when you set the `keep the screen awake` on, the screen is still locked after the preference locked time ends? Have you ever minimized the Neuron window?
> 
> I can not remember whether I have minimized the Neuron window.Because I have not see any tips about Neuron window operation before,so I do not pay close attention to this point.I have record a new screen recording before locking screen.
> 
>  2024-04-11.14.02.37.mov 
> After half an hour:
> 
>  2024-04-11.14.30.43.mov 
> [main.log](https://github.com/nervosnetwork/neuron/files/14942224/main.log)
> 
> And after half an hour I will offer a sync process picture.

To test this PR, I suggest setting the `Computer`'s locked time to 1 minute, but not half an hour, then we will test this PR quickly but not need to wait for half an hour.

**Comment by silySuper** [comment]
> > > > yes
> > > 
> > > 
> > > So, the actual problem is when you set the `keep the screen awake` on, the screen is still locked after the preference locked time ends? Have you ever minimized the Neuron window?
> > 
> > 
> > I can not remember whether I have minimized the Neuron window.Because I have not see any tips about Neuron window operation before,so I do not pay close attention to this point.I have record a new screen recording before locking screen.
> > 2024-04-11.14.02.37.mov
> > After half an hour:
> > 2024-04-11.14.30.43.mov
> > [main.log](https://github.com/nervosnetwork/neuron/files/14942224/main.log)
> > And after half an hour I will offer a sync process picture.
> 
> To test this PR, I suggest setting the `Computer`'s locked time to 1 minute, but not half an hour, then we will test this PR quickly but not need to wait for half an hour.

1 minute can not see sync process clearly,sometimes sync can not sync quickly before.

**Review by silySuper — APPROVED** [review]

**Comment by yanguoyu** [comment]
> > > > > yes
> > > > 
> > > > 
> > > > So, the actual problem is when you set the `keep the screen awake` on, the screen is still locked after the preference locked time ends? Have you ever minimized the Neuron window?
> > > 
> > > 
> > > I can not remember whether I have minimized the Neuron window.Because I have not see any tips about Neuron window operation before,so I do not pay close attention to this point.I have record a new screen recording before locking screen.
> > > 2024-04-11.14.02.37.mov
> > > After half an hour:
> > > 2024-04-11.14.30.43.mov
> > > [main.log](https://github.com/nervosnetwork/neuron/files/14942224/main.log)
> > > And after half an hour I will offer a sync process picture.
> > 
> > 
> > To test this PR, I suggest setting the `Computer`'s locked time to 1 minute, but not half an hour, then we will test this PR quickly but not need to wait for half an hour.
> 
> 1 minute can not see sync process clearly,sometimes sync can not sync quickly before.

I can not get the point, this PR is to avoid the screen locked when the switch opens.
IMO, to test the sync with the lock screen, it's unnecessary to use this package. Another package can also test that case.

**Comment by silySuper** [comment]
I click wrong

**Comment by silySuper** [comment]
> > > > > > yes
> > > > > 
> > > > > 
> > > > > So, the actual problem is when you set the `keep the screen awake` on, the screen is still locked after the preference locked time ends? Have you ever minimized the Neuron window?
> > > > 
> > > > 
> > > > I can not remember whether I have minimized the Neuron window.Because I have not see any tips about Neuron window operation before,so I do not pay close attention to this point.I have record a new screen recording before locking screen.
> > > > 2024-04-11.14.02.37.mov
> > > > After half an hour:
> > > > 2024-04-11.14.30.43.mov
> > > > [main.log](https://github.com/nervosnetwork/neuron/files/14942224/main.log)
> > > > And after half an hour I will offer a sync process picture.
> > > 
> > > 
> > > To test this PR, I suggest setting the `Computer`'s locked time to 1 minute, but not half an hour, then we will test this PR quickly but not need to wait for half an hour.
> > 
> > 
> > 1 minute can not see sync process clearly,sometimes sync can not sync quickly before.
> 
> I can not get the point, this PR is to avoid the screen locked when the switch opens. IMO, to test the sync with the lock screen, it's unnecessary to use this package. Another package can also test that case.

what does this mean?

**Review by silySuper — APPROVED** [review]

**Pull request merged**